Implement precomputed pubkey for C2/device. Resolves #19.

diagprov opened the pull request:
Logically, this implementation will derive an in-memory shared secret
for the C2/device channel under the following conditions:

 * On load from storage.
 * On setting the C2 public key if the device key is already set.
 * On setting the device secret key if the C2 public key is set.

 For these purposes, "already set" is interpreted as "not all zero
 bytes" as a best-effort interpretation.

 Storages are responsible for implementing this behaviour by calling the
 appropriate e4c functions (now added).

diagprov committed:
Implement precomputed pubkey for C2/device. Resolves #19.

Logically, this implementation will derive an in-memory shared secret
for the C2/device channel under the following conditions:

 * On load from storage.
 * On setting the C2 public key if the device key is already set.
 * On setting the device secret key if the C2 public key is set.

 For these purposes, "already set" is interpreted as "not all zero
 bytes" as a best-effort interpretation.

 Storages are responsible for implementing this behaviour by calling the
 appropriate e4c functions (now added).

daeMOn63 commented on src/e4c_pk_store_mem.c:
```diff
@@ -75,8 +75,13 @@ int e4c_set_id(e4storage *store, const uint8_t *id)
+    size_t c2keynotempty = 0;
     memmove(store->privkey, key, sizeof(store->privkey));
     e4c_sync(store);
+    c2keynotempty = zerocheck(store->c2key, sizeof(store->c2key));
```
Could it be better to return an error when `c2keynotempty === false` ? As far as I understand it we'll update the sk but not the sharedkey if the c2key is zeroed somehow.

diagprov replied:
The problem is a chicken and egg one. An end user might set the secret key first, or they might set the c2 key first. In this case, they're setting the device secret key - if we don't have the c2 public key we can't derive a shared point at this stage. If we do, we can and that's what this checks for.

So yes, we'll update the sharedkey if the c2 public key is available, which I'm taking as "non-zero" since we explicitly init all fields to zero at the top.

daeMOn63 commented on src/e4c_pk_store_mem.c:
```diff
@@ -299,16 +304,39 @@ int e4c_reset_devices(e4storage* store)
     memcpy(store->c2key, key, E4_PK_X25519_PUBKEY_LEN);
     e4c_sync(store);
-    return 0;
+    devicekeynotempty = zerocheck(store->privkey, sizeof(store->privkey));
```
Same question than `e4c_set_idseckey`, an empty devicekey sounds like an error to me

diagprov replied:
Same comment as above, more or less. This is the inverse case: the c2 public key is being set, but we don't know if the device privkey has been set yet (maybe not). So we check. If it is all zero, we don't do any derivation.

Note that if you do e4c_load() from the backing storage, this also computes the shared key because it is assumed that if the data has been written to the file, then at that point it is probably OK to do so. I should probably add zero-checks there as well.

diagprov replied:
Where we should error, actually, is if the shared point is all zero. I'll add this.

diagprov replied:
This wasn't necessary, we already error if the shared point is zero. What we did need to check was the load issue, now done.

diagprov committed:
Update load behaviour to only derive public key if both keys are present.

This commit fixes the corner case where a consuming service, for
example, sets the c2 public key but not the device secret key, then
syncs to storage, then reloads from storage. Previously a shared key
would be derived under these conditions. Since all zeros will be written
out the loaded key material will also be all zero so we check for this
to ensure the key is initialized.
